From 4f186bf63efef2d235262840a72055608fdf5dce Mon Sep 17 00:00:00 2001
From: Joshua Cranmer <joshua.cranmer@intel.com>
Date: Mon, 9 May 2022 15:40:22 -0400
Subject: [PATCH 039/172] Remove a few more getPointerElementType() calls.

The change in getScalarOrArrayConstantInt() is not actually exercised by the
current test suite, and it may be impossible to exercise it given that the
caller of the function relies on the parameter being a GEP of array type.
However, this change should correctly deduce the type were it a correct call.

The call in SPIRVRegularizeLLVM is of course unnecessary with opaque pointers:
addrspacecast can never change the pointee type of a pointer type in such
scenarios.

(cherry picked from commit 4b415cb78076690b3c8ea513a2843a6c7f466e9c)
---
 lib/SPIRV/SPIRVRegularizeLLVM.cpp | 9 ++++-----
 lib/SPIRV/SPIRVUtil.cpp           | 8 ++++++--
 2 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/lib/SPIRV/SPIRVRegularizeLLVM.cpp b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
index ff985076..8e86b9aa 100644
--- a/lib/SPIRV/SPIRVRegularizeLLVM.cpp
+++ b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
@@ -637,12 +637,11 @@ bool SPIRVRegularizeLLVMBase::regularize() {
         // Add an additional bitcast in case address space cast also changes
         // pointer element type.
         if (auto *ASCast = dyn_cast<AddrSpaceCastInst>(&II)) {
-          Type *DestTy = ASCast->getDestTy();
-          Type *SrcTy = ASCast->getSrcTy();
-          if (DestTy->getPointerElementType() !=
-              SrcTy->getPointerElementType()) {
+          PointerType *DestTy = cast<PointerType>(ASCast->getDestTy());
+          PointerType *SrcTy = cast<PointerType>(ASCast->getSrcTy());
+          if (!DestTy->hasSameElementTypeAs(SrcTy)) {
             PointerType *InterTy = PointerType::getWithSamePointeeType(
-                cast<PointerType>(DestTy), SrcTy->getPointerAddressSpace());
+                DestTy, SrcTy->getPointerAddressSpace());
             BitCastInst *NewBCast = new BitCastInst(
                 ASCast->getPointerOperand(), InterTy, /*NameStr=*/"", ASCast);
             AddrSpaceCastInst *NewASCast =
diff --git a/lib/SPIRV/SPIRVUtil.cpp b/lib/SPIRV/SPIRVUtil.cpp
index 93e3f1b2..c331b851 100644
--- a/lib/SPIRV/SPIRVUtil.cpp
+++ b/lib/SPIRV/SPIRVUtil.cpp
@@ -1233,8 +1233,12 @@ Value *getScalarOrArrayConstantInt(Instruction *Pos, Type *T, unsigned Len,
     assert(Len == 1 && "Invalid length");
     return ConstantInt::get(IT, V, IsSigned);
   }
-  if (auto PT = dyn_cast<PointerType>(T)) {
-    auto ET = PT->getPointerElementType();
+  if (isa<PointerType>(T)) {
+    unsigned PointerSize =
+        Pos->getModule()->getDataLayout().getPointerTypeSizeInBits(T);
+    auto *ET = Type::getIntNTy(T->getContext(), PointerSize);
+    assert(cast<PointerType>(T)->isOpaqueOrPointeeTypeMatches(ET) &&
+           "Pointer-to-non-size_t arguments are not valid for this call");
     auto AT = ArrayType::get(ET, Len);
     std::vector<Constant *> EV(Len, ConstantInt::get(ET, V, IsSigned));
     auto CA = ConstantArray::get(AT, EV);
-- 
2.20.1

